pkg/kvstore/allocator/cache.go

```diff
@@ -0,0 +1,280 @@
+		return NoID
+	}
+
+	return ID(id)
+}
+
+// start requests a LIST operation from the kvstore and starts watching the
+// prefix in a go subroutine.
+func (c *cache) start(a *Allocator) waitChan {
+	listDone := make(waitChan)
+
+	logger := c.getLogger()
+	logger.Info("Starting to watch allocation changes")
+
+	c.mutex.Lock()
+
+	// start with a fresh nextCache
+	c.nextCache = idMap{}
+	c.nextKeyCache = keyMap{}
+	c.mutex.Unlock()
+
+	c.stopWatchWg.Add(1)
+
+	go func() {
+		watcher := c.backend.ListAndWatch(c.prefix, c.prefix, 512)
+
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					goto abort
+				}
+				if event.Typ == kvstore.EventTypeListDone {
+					c.mutex.Lock()
+					// nextCache is valid, point the live cache to it
+					c.cache = c.nextCache
+					c.keyCache = c.nextKeyCache
+					c.mutex.Unlock()
+
+					// report that the list operation has
+					// been completed and the allocator is
+					// ready to use
+					close(listDone)
+					continue
+				}
+
+				id := c.keyToID(event.Key, c.deleteInvalidPrefixes)
+				if id != 0 {
+					c.mutex.Lock()
+
+					var key AllocatorKey
+
+					if len(event.Value) > 0 {
+						var err error
+						key, err = a.keyType.PutKey(string(event.Value))
+						if err != nil {
+							logger.WithError(err).WithField(fieldKey, event.Value).
+								Warning("Unable to unmarshal allocator key")
+						}
+					}
+					debugFields := logger.WithFields(logrus.Fields{fieldKey: key, fieldID: id})
+
+					switch event.Typ {
+					case kvstore.EventTypeCreate:
+						kvstore.Trace("Adding id to cache", nil, debugFields.Data)
+						c.nextCache[id] = key
+						if key != nil {
+							c.nextKeyCache[key.GetKey()] = id
+						}
+
+					case kvstore.EventTypeModify:
+						kvstore.Trace("Modifying id in cache", nil, debugFields.Data)
+						if k, ok := c.nextCache[id]; ok {
+							delete(c.nextKeyCache, k.GetKey())
+						}
+
+						c.nextCache[id] = key
+						if key != nil {
+							c.nextKeyCache[key.GetKey()] = id
+						}
+
+					case kvstore.EventTypeDelete:
+						kvstore.Trace("Removing id from cache", nil, debugFields.Data)
+
+						if k, ok := c.nextCache[id]; ok {
+							delete(c.nextKeyCache, k.GetKey())
+						}
+
+						delete(c.nextCache, id)
+					}
+					c.mutex.Unlock()
+
+					if a.events != nil {
+						a.events <- AllocatorEvent{
+							Typ: event.Typ,
+							ID:  ID(id),
+							Key: key,
+						}
+					}
+				}
+
+			case <-c.stopChan:
+				goto abort
+			}
+		}
+
+	abort:
+		watcher.Stop()
+		// Signal that watcher is done
+		c.stopWatchWg.Done()
+	}()
+
+	return listDone
+}
+
+func (c *cache) startAndWait(a *Allocator) error {
+	listDone := c.start(a)
+
+	// Wait for watcher to be started and for list operation to succeed
+	select {
+	case <-listDone:
+	case <-time.After(listTimeout):
+		return fmt.Errorf("Time out while waiting for list operation to complete")
+	}
+
+	return nil
+}
+
+func (c *cache) stop() {
+	select {
+	case c.stopChan <- true:
```

Comment:
You need to use `close(c.stopChan)` instead of `c.stopChan <- true`, if you call `stop()` with 2 or more `allocators` only one of them will receive the `true`. If you use `close()` then all of them will receive the event to stop. Be aware that, once `stop()` is called, you can't call `stop()` again otherwise you will get a panic for trying to close a channel 2 or more times.

Reply:
I don't understand "with 2 more or more allocators". A cache is always associated with exactly one allocator and `stop()` is called once.

The problem with using `close()` is that we would need to re-create the channel when the watcher is restarted. It gets a lot more complicated because I don't want to re-create the channel in `stop()` itself.

Reply:
Discussed ofline. Everything looks good